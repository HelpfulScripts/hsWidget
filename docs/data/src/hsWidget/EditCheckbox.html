
                <html><style>body { overflow:hidden;}
            h1 { font-family: Arial, sans-serif; font-size: 24px; color: #44a; }
            p { margin:0; padding:0; }
            br  { margin:0; padding:0; }
            .line { 
                white-space: nowrap;
                height:16px; 
            }
            .line>span { 
                display:inline-block; background-color:#eef; height:100%; 
                margin: 0 5px 0 0; padding-right: 5px; color:#999;   
            }
            comment {color: #080;} module {color: #804;} 
            quote {color: #008;} comment>quote {color: #080;}
            .listing { margin: 10px; border: 1px solid #ccc; 
                    font-family: SFMono-Regular, Consolas, 'Liberation Mono', Menlo, Courier, monospace;
                    font-size: 14px;  
                    overflow:scroll;
                    height:90%;
            }
            code { padding: 5px 0;}</style></html><body>
                <h1>EditCheckbox.ts</h1>
                <div class='listing'><code><p id=1 class="line"><span>&nbsp;&nbsp;&nbsp;1</span><comment>/**</comment></p>
<p id=2 class="line"><span>&nbsp;&nbsp;&nbsp;2</span><comment>&nbsp;*&nbsp;#&nbsp;EditCheckbox</comment></p>
<p id=3 class="line"><span>&nbsp;&nbsp;&nbsp;3</span><comment>&nbsp;*&nbsp;Provides&nbsp;a&nbsp;checkbox</comment></p>
<p id=4 class="line"><span>&nbsp;&nbsp;&nbsp;4</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=5 class="line"><span>&nbsp;&nbsp;&nbsp;5</span><comment>&nbsp;*&nbsp;###&nbsp;Attributes:</comment></p>
<p id=6 class="line"><span>&nbsp;&nbsp;&nbsp;6</span><comment>&nbsp;*&nbsp;-&nbsp;**checked**:&nbsp;`boolean`&nbsp;if&nbsp;true,&nbsp;the&nbsp;checkbox&nbsp;is&nbsp;initially&nbsp;checked</comment></p>
<p id=7 class="line"><span>&nbsp;&nbsp;&nbsp;7</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=8 class="line"><span>&nbsp;&nbsp;&nbsp;8</span><comment>&nbsp;*&nbsp;###&nbsp;Example</comment></p>
<p id=9 class="line"><span>&nbsp;&nbsp;&nbsp;9</span><comment>&nbsp;*&nbsp;<example></comment></p>
<p id=10 class="line"><span>&nbsp;&nbsp;10</span><comment>&nbsp;*&nbsp;<file&nbsp;name=<quote>'script.js'</quote>>&nbsp;</comment></p>
<p id=11 class="line"><span>&nbsp;&nbsp;11</span><comment>&nbsp;*&nbsp;let&nbsp;selected&nbsp;=&nbsp;<quote>'none'</quote>;</comment></p>
<p id=12 class="line"><span>&nbsp;&nbsp;12</span><comment>&nbsp;*&nbsp;m.mount(root,&nbsp;{view:&nbsp;()&nbsp;=>&nbsp;m(<quote>'div'</quote>,&nbsp;[</comment></p>
<p id=13 class="line"><span>&nbsp;&nbsp;13</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;m(<quote>'h3'</quote>,&nbsp;<quote>'Checkbox&nbsp;Example'</quote>),</comment></p>
<p id=14 class="line"><span>&nbsp;&nbsp;14</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;m(<quote>'div'</quote>,&nbsp;[</comment></p>
<p id=15 class="line"><span>&nbsp;&nbsp;15</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;m(<quote>'span'</quote>,&nbsp;`current&nbsp;content:&nbsp;<quote>'${selected}'</quote>`),</comment></p>
<p id=16 class="line"><span>&nbsp;&nbsp;16</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;m(hsWidget.Editheckbox,&nbsp;{</comment></p>
<p id=17 class="line"><span>&nbsp;&nbsp;17</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;css:&nbsp;<quote>'.mySelect'</quote>,</comment></p>
<p id=18 class="line"><span>&nbsp;&nbsp;18</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;update:&nbsp;newValue&nbsp;=>&nbsp;selected&nbsp;=&nbsp;newValue</comment></p>
<p id=19 class="line"><span>&nbsp;&nbsp;19</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;})</comment></p>
<p id=20 class="line"><span>&nbsp;&nbsp;20</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;])</comment></p>
<p id=21 class="line"><span>&nbsp;&nbsp;21</span><comment>&nbsp;*&nbsp;])});</comment></p>
<p id=22 class="line"><span>&nbsp;&nbsp;22</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=23 class="line"><span>&nbsp;&nbsp;23</span><comment>&nbsp;*&nbsp;</file></comment></p>
<p id=24 class="line"><span>&nbsp;&nbsp;24</span><comment>&nbsp;*&nbsp;<file&nbsp;name=<quote>'style.css'</quote>></comment></p>
<p id=25 class="line"><span>&nbsp;&nbsp;25</span><comment>&nbsp;*&nbsp;.mySelect&nbsp;{&nbsp;&nbsp;</comment></p>
<p id=26 class="line"><span>&nbsp;&nbsp;26</span><comment>&nbsp;*&nbsp;}</comment></p>
<p id=27 class="line"><span>&nbsp;&nbsp;27</span><comment>&nbsp;*&nbsp;</file></comment></p>
<p id=28 class="line"><span>&nbsp;&nbsp;28</span><comment>&nbsp;*&nbsp;</example></comment></p>
<p id=29 class="line"><span>&nbsp;&nbsp;29</span><comment>&nbsp;*/</comment></p>
<p id=30 class="line"><span>&nbsp;&nbsp;30</span></p>
<p id=31 class="line"><span>&nbsp;&nbsp;31</span><comment>/**&nbsp;*/</comment></p>
<p id=32 class="line"><span>&nbsp;&nbsp;32</span>import&nbsp;m&nbsp;from&nbsp;<quote>"mithril"</quote>;</p>
<p id=33 class="line"><span>&nbsp;&nbsp;33</span>type&nbsp;Vnode&nbsp;=&nbsp;m.Vnode<any,&nbsp;any>;</p>
<p id=34 class="line"><span>&nbsp;&nbsp;34</span>import&nbsp;{&nbsp;Log&nbsp;}&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;from&nbsp;<quote>'hsutil'</quote>;&nbsp;&nbsp;const&nbsp;log&nbsp;=&nbsp;new&nbsp;Log(<quote>'EditSelect'</quote>);</p>
<p id=35 class="line"><span>&nbsp;&nbsp;35</span></p>
<p id=36 class="line"><span>&nbsp;&nbsp;36</span></p>
<p id=37 class="line"><span>&nbsp;&nbsp;37</span>export&nbsp;class&nbsp;EditCheckbox&nbsp;{</p>
<p id=38 class="line"><span>&nbsp;&nbsp;38</span>&nbsp;&nbsp;&nbsp;&nbsp;update:(checked:boolean)&nbsp;=>&nbsp;void;</p>
<p id=39 class="line"><span>&nbsp;&nbsp;39</span>&nbsp;&nbsp;&nbsp;&nbsp;checked:boolean;</p>
<p id=40 class="line"><span>&nbsp;&nbsp;40</span></p>
<p id=41 class="line"><span>&nbsp;&nbsp;41</span>&nbsp;&nbsp;&nbsp;&nbsp;click()&nbsp;{</p>
<p id=42 class="line"><span>&nbsp;&nbsp;42</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.checked&nbsp;=&nbsp;!this.checked;</p>
<p id=43 class="line"><span>&nbsp;&nbsp;43</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.update(this.checked);</p>
<p id=44 class="line"><span>&nbsp;&nbsp;44</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=45 class="line"><span>&nbsp;&nbsp;45</span></p>
<p id=46 class="line"><span>&nbsp;&nbsp;46</span>&nbsp;&nbsp;&nbsp;&nbsp;oninit(node:Vnode)&nbsp;{</p>
<p id=47 class="line"><span>&nbsp;&nbsp;47</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.checked&nbsp;=&nbsp;node.attrs.checked;</p>
<p id=48 class="line"><span>&nbsp;&nbsp;48</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.update&nbsp;&nbsp;=&nbsp;node.attrs.update;</p>
<p id=49 class="line"><span>&nbsp;&nbsp;49</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=50 class="line"><span>&nbsp;&nbsp;50</span></p>
<p id=51 class="line"><span>&nbsp;&nbsp;51</span>&nbsp;&nbsp;&nbsp;&nbsp;view(node:Vnode)&nbsp;{</p>
<p id=52 class="line"><span>&nbsp;&nbsp;52</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;const&nbsp;css&nbsp;=&nbsp;node.attrs.css&nbsp;||&nbsp;<quote>''</quote>;</p>
<p id=53 class="line"><span>&nbsp;&nbsp;53</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return&nbsp;m(`input.hsedit_checkbox${css}`,&nbsp;{&nbsp;</p>
<p id=54 class="line"><span>&nbsp;&nbsp;54</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;type:&nbsp;<quote>'checkbox'</quote>,</p>
<p id=55 class="line"><span>&nbsp;&nbsp;55</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;checked:&nbsp;this.checked,</p>
<p id=56 class="line"><span>&nbsp;&nbsp;56</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;onclick:&nbsp;this.click.bind(this)</p>
<p id=57 class="line"><span>&nbsp;&nbsp;57</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;},<quote>''</quote>);</p>
<p id=58 class="line"><span>&nbsp;&nbsp;58</span></p>
<p id=59 class="line"><span>&nbsp;&nbsp;59</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=60 class="line"><span>&nbsp;&nbsp;60</span>}</p></code></div>
                </body>
            