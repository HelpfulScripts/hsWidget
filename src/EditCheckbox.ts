/**
 * # EditCheckbox
 * Provides a checkbox
 * 
 * ### Profile
 * invoked as `m(EditCheckbox, { <EditCheckbox> }, <content>)`.
 * 
 * See {@link EditCheckbox.EditCheckboxAttrs EditCheckbox}
 * `<content>`: an optional string that will be displayed as the checkbox label; defaults to ''
 * 
 * ### Example
 * <example>
 * <file name='script.js'> 
 * let selected = false;
 * m.mount(root, {view: () => m('div', [
 *    m('h3', 'Checkbox Example'),
 *    m('div', [
 *      m(hsWidget.EditCheckbox, {
 *          class: 'mySelect',
 *          update: newValue => selected = newValue
 *      }, `${selected}: I am ${selected?'' : 'not '}a robot`)
 *    ])
 * ])});
 * 
 * </file>
 * <file name='style.css'>
 * .mySelect {  
 * }
 * </file>
 * </example>
 */

/** */
import m                from "mithril";
import { Log }          from 'hsutil'; const log = new Log('EditSelect');
import { Vnode }        from "./Widget";
import { WidgetAttrs }  from "./Widget";
import { Widget }       from "./Widget";


export interface EditCheckboxAttrs extends WidgetAttrs {
    /** if `true`, the checkbox will initially be checked. Defaults to `false` */
    initial?: boolean;
    update: (checked:boolean) => void;
}


export class EditCheckbox extends Widget {
    /** unique ID for use in `label for` */
    id: string;
    /** `true` if the checkbox is checked. */
    checked: boolean;

    toggleCheckbox: ()=>void;

    oninit(node:Vnode<EditCheckboxAttrs, this>) {
        node.state.id = node.attrs.id || `hsCheckbox${Math.floor(Math.random()*1000000)}`;
        node.state.checked = node.attrs.initial || false;
        node.state.toggleCheckbox = () => {
            node.state.checked = !node.state.checked;
            node.attrs.update(node.state.checked);
        }
    }

    view(node:Vnode<EditCheckboxAttrs, this>) {
        return m('.hs_edit_checkbox', this.attrs(node.attrs, <any>{ 
        }), [
            m(`input`, {
                id:node.state.id,
                type: 'checkbox',
                checked: node.state.checked,
                onclick: node.state.toggleCheckbox
            }),
            m('label', {for:node.state.id}, node.children||'')
        ])
    }
}