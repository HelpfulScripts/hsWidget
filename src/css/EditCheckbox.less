.hs_edit_checkbox {
    height: 30px;
    input {
        background-color: #eee;
        border: 1px solid #88c;
        width: 30px;
        height: 100%;
        border-radius: 15px;
    }
    label {
        display: inline-block;
        vertical-align: middle;
        height:100%;
    }
}