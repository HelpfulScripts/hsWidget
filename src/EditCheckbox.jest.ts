import { m }      from 'hslayout';
import { EditCheckbox }    from './EditCheckbox';
const root = window.document.createElement("div");

let selected = '';

describe('EditCheckbox', () => {
    beforeAll(()=>{
        m.mount(root, { view: () => m(EditCheckbox, {
            css: '.mySelect',
            update: (newValue:string) => selected = newValue
        })});
    });

    it('should match snapshot', () => {
        expect(root).toMatchSnapshot();
    });
});