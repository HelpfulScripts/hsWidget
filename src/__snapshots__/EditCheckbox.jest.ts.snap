// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditCheckbox should match snapshot 1`] = `
<div>
  <input
    class="hsedit_checkbox mySelect"
    type="checkbox"
  >
    
  </input>
</div>
`;
